Outliers := smth we that has never been predicted
- Look at distribution; why are the outliers outliers? 


backing common sense (e.g. from reputable sources)



# Import all libraries here

In [1]:
import pandas as pd

In [2]:
ls ../data/raw/all_rental_data

All_Houses_Scraped.csv  All_Houses_Scraped.csv:Zone.Identifier


# Import all rental data

In [3]:
rental_raw = pd.read_csv('../data/raw/All_Houses_Scraped.csv')

In [4]:
rental_raw.head()

Unnamed: 0  \
0  https://www.domain.com.au/15-fintonia-street-h...   
1  https://www.domain.com.au/41-bishop-street-oak...   
2  https://www.domain.com.au/109-6-dalgety-street...   
3  https://www.domain.com.au/16-30-swindon-road-o...   
4  https://www.domain.com.au/1-14-bletchley-road-...   

                  coordinates cost_text                     desc  \
0  [-37.9032708, 145.0770553]      $680  Joseph Chatziconstantis   
1   [-37.9077745, 145.092738]   $650.00        James Drakopoulos   
2  [-37.8922004, 145.0899378]      $500               Nigel Chee   
3  [-37.8972564, 145.0871089]      $570       Matthew Swinnerton   
4    [-37.894307, 145.082706]      $475                 Jason Gu   

                                      name                               rooms  
0   15 Fintonia Street Hughesdale VIC 3166  ['3 Beds', '2 Baths', '2 Parking']  
1       41 Bishop Street Oakleigh VIC 3166  ['3 Beds', '2 Baths', '2 Parking']  
2   109/6 Dalgety Street Oakleigh VIC 3166    ['1 Bed', '1 Bath', '1 Parking']  
3     16/30 Swindon Road Oakleigh VIC 3166  ['3 Beds', '2 Baths', '2 Parking']  
4  1/14 Bletchley Road Hughesdale VIC 3166   ['2 Beds', '1 Bath', '1 Parking']

# Obtain all values of <code>rental_raw['cost_text']</code> that have numbers in it
- treat non-numeric values as missing values

In [5]:
NUMBERS_SET = set(list('0123456789'))
are_costs_numeric = rental_raw['cost_text'].map(
    #returns True if there's at least 1 of 0123456789 in s
    lambda s: bool(set(list(s)) & NUMBERS_SET)
    )

Check if the prices contain "worded" numbers (1 - one, etc)

In [6]:
WORD_NUMS = ['one','two','three','four','five','six','seven','eight','nine','ten','zero']

for num_word in WORD_NUMS:
    print(rental_raw['cost_text'].str.lower().str.contains(num_word).sum())

2
0
0
0
0
0
0
0
0
5
0


In [7]:
rental_raw['cost_text'].str.lower()[
    rental_raw['cost_text'].str.lower().str.contains('ten')
]

543        $495 per week including garden maintenance
2482     $480 per week (including garden maintenance)
8131        $560 per week-includes garden maintenance
13317     $1,000 per week- including pool maintenance
14552       $1600 per week including pool maintenance
Name: cost_text, dtype: object

In [8]:
rental_raw['cost_text'][~are_costs_numeric].shape[0] / rental_raw['cost_text'].shape[0]

0.02242637125101324

Only 2% instances contain non-numerics

# <font color='green'> Preproc 1: Remove instances with non-numeric values of <code>cost_text</code></font>

Copy <code>rental_raw</code> to another variable

In [9]:
rental_raw_data = rental_raw.copy()

Do as per the subheading above

In [10]:
rental_raw_data = rental_raw_data[are_costs_numeric]

# <font color='green'> Preproc 2: convert all costs values to lower cases </font>

In [11]:
rental_raw_data['cost_text'] = rental_raw_data['cost_text'].str.lower()

# Inspect the modified data

Obtain all instances that contain 'w'
- idea is to look which kind of instances that have the prices written as (pw, per week, etc)

In [12]:
rental_raw_data[rental_raw_data['cost_text'].str.contains('w')]

Unnamed: 0  \
5      https://www.domain.com.au/26-austin-street-hug...   
10     https://www.domain.com.au/1-61-macrina-street-...   
27     https://www.domain.com.au/501-19-21-hanover-st...   
29     https://www.domain.com.au/30-oxford-st-oakleig...   
30     https://www.domain.com.au/1-20-dover-st-oaklei...   
...                                                  ...   
14790  https://www.domain.com.au/34a-st-albans-street...   
14791  https://www.domain.com.au/1-hilsea-court-mount...   
14794  https://www.domain.com.au/46-virginia-street-m...   
14796  https://www.domain.com.au/40-anthony-dr-mount-...   
14801  https://www.domain.com.au/1-30-ailsa-street-ma...   

                      coordinates               cost_text  \
5      [-37.9068017, 145.0786493]           $550 per week   
10     [-37.9029684, 145.1135954]           $430 per week   
27      [-37.9010428, 145.091833]           $370 per week   
29     [-37.9012582, 145.0929966]           $1,150 weekly   
30        [-37.90993, 145.121535]             $590 weekly   
...                           ...                     ...   
14790  [-37.8736601, 145.1342532]           $950 per week   
14791  [-37.8591721, 145.1332939]           $460 per week   
14794   [-37.878088, 145.1249743]           $640 per week   
14796  [-37.8900914, 145.1353848]             $525 weekly   
14801  [-37.0551267, 146.0929647]  $440p/w ($1911.90 pcm)   

                                                    desc  \
5                                          Nina Del Toro   
10                                          Leasing Team   
27     This stunning apartment is perfectly located t...   
29     Perfectly located in central Oakleigh within w...   
30     This beautifully double storey townhouse is in...   
...                                                  ...   
14790           PROPERTY ID: 112553 (quote when calling)   
14791   class="css-dxogle">* Unverified feature<svg a...   
14794   class="css-dxogle">* Unverified feature<svg a...   
14796   class="css-dxogle">* Unverified feature<svg a...   
14801   class="css-dxogle">* Unverified feature<svg a...   

                                               name  \
5              26 Austin Street Hughesdale VIC 3166   
10       1/61 Macrina Street Oakleigh East VIC 3166   
27       501/19-21 Hanover Street Oakleigh VIC 3166   
29                   30 Oxford St Oakleigh VIC 3166   
30             1/20 Dover St Oakleigh East VIC 3166   
...                                             ...   
14790  34a St Albans Street Mount Waverley VIC 3149   
14791        1 Hilsea Court Mount Waverley VIC 3149   
14794    46 Virginia Street Mount Waverley VIC 3149   
14796         40 Anthony Dr Mount Waverley VIC 3149   
14801          1/30 Ailsa Street Mansfield VIC 3722   

                                    rooms  
5       ['3 Beds', '1 Bath', '2 Parking']  
10      ['3 Beds', '1 Bath', '1 Parking']  
27       ['1 Bed', '1 Bath', '1 Parking']  
29      ['4 Beds', '1 Bath', '4 Parking']  
30     ['4 Beds', '3 Baths', '2 Parking']  
...                                   ...  
14790  ['4 Beds', '2 Baths', '2 Parking']  
14791   ['4 Beds', '1 Bath', '2 Parking']  
14794  ['3 Beds', '2 Baths', '2 Parking']  
14796  ['4 Beds', '2 Baths', '2 Parking']  
14801   ['2 Beds', '1 Bath', '3 Parking']  

[8873 rows x 6 columns]

In [13]:
sorted(
    list(rental_raw_data[
    rental_raw_data['cost_text'].str.contains('w')
].\
    cost_text.\
        unique()),
    key= lambda s: len(s),
    reverse=True
)

['$325 per week until april 2022. increase to $335 per week a',
 'leased - more properties wanted call andrew 0407 060 344',
 '$550 per week until 31 jan 2023, after $600 per week',
 '$510pw for a 12 month lease or $530pw for 6 months',
 '$195 per week - then $245 per week from 10/02/2023',
 'apply now - studio from $250/w & 2 bed from $385/w',
 '1 weeks free rent then $165 per week all inclusive',
 '$220.00 pw- (rent increase to $285 pw from 25/1/23',
 'clean, bright, modern private studio from $290pw',
 '$580pw - great views and nbn connection paid for',
 '$240 pw until 20/01/2023 then increase to $320pw',
 'your luxurious lifestyle awaits | $950 per week',
 '$650/w, brand new unit with stunning city views',
 '$695 per week fully furnished short term rental',
 '$750 per week incl. gardening or $730 per week',
 '$490pw / $2,129 pcm - 6 month lease available',
 '$525pw - art precinct views plus storage cage',
 '5 mins to tram & shops & school,$880/per week',
 '$450 pw - save time view 

(Try to) Find all instances such that its cost is defined per week

Normally, format would be something like <code>PRICE__w</code> where <code>__</code> could be any character and <code>w</code> could be either <code>week, weekly</code> etc.

In [14]:
rental_raw_data['cost_text'].str.findall(r'\d+.*w')[
    rental_raw_data['cost_text'].str.findall(r'\d+.*w').map(lambda x: bool(x))
]

5        [550 per w]
10       [430 per w]
27       [370 per w]
29         [1,150 w]
30           [590 w]
            ...     
14790    [950 per w]
14791    [460 per w]
14794    [640 per w]
14796        [525 w]
14801       [440p/w]
Name: cost_text, Length: 8872, dtype: object

Maybe we can make sure that the <code>w</code> corresponds to <code>week</code>. One way to do this is to print out the rest of the values after <code>w</code>

In [15]:
syntax = r'\d+.*w.*'
test_rental_raw_cost_per_week = rental_raw_data['cost_text'].str.findall(syntax)[
    rental_raw_data['cost_text'].str.findall(syntax).map(lambda x: bool(x))    
]
test_rental_raw_cost_per_week

5                 [550 per week]
10                [430 per week]
27                [370 per week]
29                [1,150 weekly]
30                  [590 weekly]
                  ...           
14790             [950 per week]
14791             [460 per week]
14794             [640 per week]
14796               [525 weekly]
14801    [440p/w ($1911.90 pcm)]
Name: cost_text, Length: 8872, dtype: object

# Extracting Rent Per Week or Other Rents Containing <code>'w'</code>
One way to do this is that find all instances for which <code>cost_text</code> contains <code>numbers__w__</code>, where: 

- <code>__</code> after <code>numbers</code> is any character
- <code>w__</code> is the first word starting with <code>w</code>
also we're accounting the possibility that some may write it as <code>pw, per week, per wk</code> and etc. 


Even if the <code>w__</code> does not signify annual rent, provided that there's <code>numbers</code> that signify the rental price, we still can assume it as the weekly rent for that property (more data is better!)

Or, maybe a capture group of the characters following <code>w</code> onwards would help.

1.
- Since we want to know whether week := <code>w, week, wk</code> (or maybe not even corresponding to week), for every occurence of any word beginning with <code>w</code>, we stop if there's any whitespace or there's any number following it to inspect the unique values








In [16]:
test_rental_raw_cost_per_week

5                 [550 per week]
10                [430 per week]
27                [370 per week]
29                [1,150 weekly]
30                  [590 weekly]
                  ...           
14790             [950 per week]
14791             [460 per week]
14794             [640 per week]
14796               [525 weekly]
14801    [440p/w ($1911.90 pcm)]
Name: cost_text, Length: 8872, dtype: object

In [17]:
syntax = r'\d+.*(w[^\s\d]*)'
test_rental_raw_cost_contains_w = rental_raw_data['cost_text'].str.findall(syntax)[
    rental_raw_data['cost_text'].str.findall(syntax).map(lambda x: bool(x))    
]

#ensure that EACH of the outputs below BELONG to EXACTLY ONE instance
#print("Max length for matches in each instance: ",test_rental_raw_cost_contains_w.map(lambda x: len(x)).max())

#print all words that begin in 'w'
print(sorted(test_rental_raw_cost_contains_w.map(lambda x: x[0]).unique()))

['w', 'w!!', 'w$', 'w**', 'w,', 'w-', 'w.', 'w/', 'w/$', 'walk', 'walking', 'wanted**', 'water', 'waterfront', 'week', 'week!', 'week!!', 'week)', 'week,', 'week-', 'week-includes', 'week.', 'week/', 'week/$', 'week/fully', 'week;', 'weekly', 'weeks', 'wek', 'well', 'weston', 'wfh', 'wing', 'with', 'wk', 'wn', 'ws', 'ws!', 'wunfurnished/']


~~Since <code>re.findall</code> is greedy (meaning if there's > 1 matches the length of list retured will > 1), and that it returns exactly one match of the pattern <code>syntax</code> from the matched instances,~~

- instances containing, e.g. <code>'wunfurnished/', 'weston', 'wfh'</code> do **not** have the *weekly* rental prices explicitly listed, but that's okay because, again, we can assume that they are weekly, provided that the rental prices for such instances are matched (by the regex <code>\d+</code> in <code>syntax</code>) 

2. 

A more better refinement is to include 'p' (weekly rental prices can be written as <code>numbers per week, numbers pw, numbers p week</code> and etc)

Now, it seems like we need to add more filter to include any <code>p</code> (and cross match with the above with the reason below:)
- Example: <code>wunfurnished/</code> could be preceded by, say <code>pwunfurnished</code> (perhaps signifying per week and unfurnished)

In [18]:
syntax = r'\d+.*(p.*w[^\s\d]*)'
test_rental_raw_cost_p_w = rental_raw_data['cost_text'].str.findall(syntax)[
    rental_raw_data['cost_text'].str.findall(syntax).map(lambda x: bool(x))    
]

#ensure that EACH of the outputs below BELONG to EXACTLY ONE instance (re.findall is greedy)
#print("Max length for matches in each instance: ",test_rental_raw_cost_p_w.map(lambda x: len(x)).max())

#print all words that begin in 'w'
print(test_rental_raw_cost_p_w.map(lambda x: x[0]).unique())

['per week' 'pw' 'per week week' 'p/w' 'pw/$' 'p.w.' 'per week/$'
 'per week;' 'pw - save time view' 'per weekly' 'per week.' 'perweek'
 'per week,' 'pw/' 'pw | weston' 'pw,' 'p/week' 'per week - new'
 'pw / avail. now' 'per week/fully' 'per week/' 'p.w' 'per week with'
 'per wek' 'pw low,' 'pw low' 'pw$' 'per week!!' 'per week)' 'pwk'
 'pw with' 'pw, with' 'per week!! available now' 'parate study/ wfh'
 'pw - s/cage! white goods! views!' 'per week!! enquire now!!'
 'per week-includes' 'per wk' 'pcm 1 weeks' 'p/w/$'
 'per week including lawn mowing' 'pw reviewing'
 'pw - study! s/cage! views!' 'precinct views' 'pw - great views'
 'p/w - big house - walk' 'pw - walk' 'pw (water' 'pw, 1 week' 'pw-'
 'per week-' 'pre week' 'per week **brand new']


Now, we check if matches in 2. is contained in matches in 1. 
- If it does then we can use the instances in <code>test_rental_raw_cost_contains_w</code> that approximately contains rental prices listed *weekly*

In [19]:
set(test_rental_raw_cost_p_w.index) - set(test_rental_raw_cost_contains_w.index)

set()

In [20]:
set(test_rental_raw_cost_p_w.index) & set(test_rental_raw_cost_contains_w.index) == set(test_rental_raw_cost_p_w.index)

True

Now, we try handpick instances for which <code>cost_text</code> are on a weekly basis. 
- To do this, we look at the instances that contains <code>w</code> through the <code>test_rental_raw_cost_contains_w</code> dataframe, print out unique values (based on previous cells such unique values are not many)

In [21]:
test_rental_raw_cost_contains_w

5          [week]
10         [week]
27         [week]
29       [weekly]
30       [weekly]
           ...   
14790      [week]
14791      [week]
14794      [week]
14796    [weekly]
14801         [w]
Name: cost_text, Length: 8872, dtype: object

In [22]:
#note that each  instance has exactly one of these
print(sorted(test_rental_raw_cost_contains_w.map(lambda x: x[0]).unique()))

['w', 'w!!', 'w$', 'w**', 'w,', 'w-', 'w.', 'w/', 'w/$', 'walk', 'walking', 'wanted**', 'water', 'waterfront', 'week', 'week!', 'week!!', 'week)', 'week,', 'week-', 'week-includes', 'week.', 'week/', 'week/$', 'week/fully', 'week;', 'weekly', 'weeks', 'wek', 'well', 'weston', 'wfh', 'wing', 'with', 'wk', 'wn', 'ws', 'ws!', 'wunfurnished/']


In [23]:
to_investigate_indicators_for_weekly_cost = sorted(test_rental_raw_cost_contains_w.map(lambda x: x[0]).unique())

# NEED TO MAKE SURE THAT INSTANCES THAT CONTAIN <code>w</code> ARE ALL RECORDED WEEKLY

Important: we assumed that any instances containing <code>w</code> either belongs to rental prices that: 
1. are recorded weekly, or
2. miscalleneous reasons 

BUT only finding instances with numbers + whether it contains <code>w</code> still opens up the possibility that some of them may have <font color='red'>costs recoded annually, or other timeframes.</font>
- e.g. maybe the word containing <code>w</code> may indicate something else other than weekly rents, and maybe followed by e.g. <code>per annum</code> or some sort

To mitigate this, we gradually inspect the picked up words starting with <code>w</code> and make sure that those kinds of instances adhere to the 2 requirements above, 
- and remove the ones for which we're sure what sort of instances contaning those values (e.g. instances with its <code>cost_text</code> containing <code>week</code> signify that the cost is recorded weekly)

Firstly, it is obvious that <code>cost_text</code>s contaning the word <code>week</code> (or <code>wek</code>, which is a typo, see below) signifies that the cost is recorded weekly.

In [24]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index]

5                 $550 per week
10                $430 per week
27                $370 per week
29                $1,150 weekly
30                  $590 weekly
                  ...          
14790             $950 per week
14791             $460 per week
14794             $640 per week
14796               $525 weekly
14801    $440p/w ($1911.90 pcm)
Name: cost_text, Length: 8872, dtype: object

In [25]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('wek')
]

2731    $560 per wek
Name: cost_text, dtype: object

In [26]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('wk')
].unique()

array(['$580 wk', '$400 wk', '$500 wk', '$420 wk', '$430 wk', '$340 wk',
       '$440 wk', '$600 wk', '$3,775 wk $545 pn', '$1,050 wk',
       '$380 / wk', '$395 wk / $1712 pcm', '$390 wk / $1690 pcm',
       '$485 pwk', '$329 / wk', '$515 / wk', '$365 / wk', '$350 / wk',
       '$345 / wk', '$300 / wk', '$255 / wk', '$369 / wk', '$455 / wk',
       '$349 / wk', '$145 / wk', '280 / wk', '$285 / wk', '260 / wk',
       '$415 / wk', '$975 / wk', '$550 wk', 'from $750 per wk', '$470 wk',
       '$240 wk', '$330 wk', '$250 wk', '$345 wk', '$380 wk', '$410 wk',
       '$300 wk', '$285 wk', '$360 wk', '$370 wk', '$749 / wk', '$450 wk',
       '$350 wk pw', '$375 /wk', '$385 / wk', '$490 wk', '$475 wk',
       '$345/ wk', '$360 / wk pw', '$525 /wk', '$570 wk', '$625 / wk'],
      dtype=object)

In [27]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('wn')
].unique()

array(['no deposit required or rent2own', 'rent2own with no deposit',
       'rent2own program', 'no bond required with rent2own', 'rent2own',
       '$450 per week including lawn mowing'], dtype=object)

In [28]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('ws')
].unique()

array(['large 2bedroom with river views',
       '$650/w, brand new unit with stunning city views',
       '$425pw - s/cage! white goods! views!',
       '$620pw - study! s/cage! views!',
       '$525pw - art precinct views plus storage cage',
       '$580pw - great views and nbn connection paid for'], dtype=object)

In [29]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('well')
].unique()

array(['$730 **well furnished**'], dtype=object)

In [30]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('wing')
].unique()

array(['$450 per week including lawn mowing',
       '$485 pw reviewing applications'], dtype=object)

In [31]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('with')
].unique()

array(['rent2own with no deposit', '$380 per week with gardener',
       'no bond required with rent2own', '$350pw with furnished',
       '$450 pw, with furnished', 'large 2bedroom with river views',
       '$650/w, brand new unit with stunning city views'], dtype=object)

In [32]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('unfurnished')
].unique()

array(['$580 per week *unfurnished*', '$600 per week *unfurnished*',
       '$530 per week - unfurnished', '$450/w unfurnished',
       '$900 per week (unfurnished)', '665 per week unfurnished',
       '$500 per week *unfurnished*',
       '$550/wunfurnished/ furnished optional', '$560/w unfurnished',
       '$800 per week unfurnished', '$580 per weekly - unfurnished',
       '$700/w unfurnished', '$500/w unfurnished',
       '$450 per week *unfurnished*', '$750/w unfurnished',
       '$900/w unfurnished', '$600/w unfurnished', '$990 w unfurnished'],
      dtype=object)

In [33]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('water')
].unique()

array(['$950 **waterfront furnished apartment**',
       '$850 **waterfront spacious 3-bedrooms**',
       '$360.00pw (water included)', '$310 water & electricity included'],
      dtype=object)

In [34]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('walk')
].unique()

array(['$300p/w - big house - walk to train', '$350 pw - walk to school',
       '$620 **walking to uni mel & rmit**'], dtype=object)

In [35]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('wanted')
].unique()

array(['$500 **stable family living wanted**'], dtype=object)

In [36]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('wfh')
].unique()

array(['$375pw - separate study/ wfh'], dtype=object)

In [37]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('weston')
].unique()

array(['$425 pw | weston real estate', '$699 pw | weston real estate',
       '$499 pw | weston real estate', '$390 pw | weston real estate'],
      dtype=object)

In [38]:
to_investigate_indicators_for_weekly_cost = [indicator for indicator in to_investigate_indicators_for_weekly_cost if\
                                                    'week' not in indicator and\
                                                    'wek' not in indicator and\
                                                    'wk' not in indicator and\
                                                    'wn' not in indicator and\
                                                    'ws' not in indicator and\
                                                    'well' not in indicator and\
                                                    'wing' not in indicator and\
                                                    'with' not in indicator and \
                                                    'unfurnished' not in indicator and \
                                                    'water' not in indicator and\
                                                    'walk' not in indicator and\
                                                    'wanted' not in indicator and\
                                                    'wfh' not in indicator and\
                                                    'weston' not in indicator]          

In [39]:
print(to_investigate_indicators_for_weekly_cost)

['w', 'w!!', 'w$', 'w**', 'w,', 'w-', 'w.', 'w/', 'w/$']


I'd imagine that there's going to be a lot of instances that contain <code>w</code>. So we inspect instances with the other values first

In [40]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('w!!')
].unique()

#yes, its recorded per week

array(['$420 per week!! enquire now!!'], dtype=object)

In [41]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('w\*\*') #else * will be not be treated as a raw string
].unique()

#not specified but we're going to assume its per week

array(['$750 **level 53 high level view**'], dtype=object)

In [42]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('w,')
].unique()

#yes, its recorded per week

array(['$450pw, $1955pcm', '$1000 pw high, $650 pw low, $160 pn',
       '$1750 pw high, $1400 pw low, $250 pn',
       '$2,000 pw high, $1,200 pw low, $300 pn',
       '$1200 pw high, $600 pw low, $180 pn',
       '$2500 pw high, $1600 pw low, $380 pn',
       '$1,200 pw high, $950 pw low, $190 pn',
       '$1200 pw high, $750 pw low, $145 pn', '$450 pw, with furnished',
       '$800pw, $3476pcm', '$950pw, $4128pcm',
       '$650/w, brand new unit with stunning city views',
       '$620pw, $2694pcm', '$560pw, $2433pcm', '$330pw, $1434pcm',
       '$440pw, $1912pcm', '$580pw, fully furnished inc utilities',
       '$500pw, $2173pcm', '$180pw, 1 week free rent',
       '$680pw, $2955.00pcm'], dtype=object)

In [43]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('w-')
].unique()

#yes, its recorded per week

array(["$250pw- 'sms' 0414523506 to inspect",
       '$220.00 pw- (rent increase to $285 pw from 25/1/23'], dtype=object)

In [44]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('w\.')
].unique()

#yes, its recorded per week

array(['$400 p.w.', '$690 p.w.', '$415 p.w.', '$580 p.w.', '$610 p.w.',
       '$550 p.w.', '$440 p.w.', '$430 p.w.', '$420 p.w.', '$450 p.w.',
       '$320 p.w.', '$300 p.w.', '$500 p.w.', '$250 p.w.', '$650 p.w.',
       '$800 p.w.', '$530 p.w.', '$640 p.w.', '$560 p.w.', '$680 p.w.',
       '$620 p.w.', '$390 p.w.', '$635 p.w.', '$750 p.w.', '$465 p.w.',
       '$270 p.w.', '$370 p.w.', '$380 p.w.', '$470 p.w.', '$350 p.w.',
       '$445 p.w.', '$355 p.w.', '$330 p.w.', '$280 p.w.', '$360 p.w.',
       '$480 p.w.', '$460 p.w.', '$480p.w.',
       'available now. from $380 p/w', '$430p.w.', '$475 p.w.',
       '$525 p.w.', '$490 p.w.', '$365 p.w.', '$510 p.w.', '$410 p.w.',
       'from $185 p.w.', '$460.00 p.w.', '$520 p.w.', '$595 p.w.',
       '$600 p.w.', '$660 p.w.', '$180 p.w.', '$485 p.w.', '$495 p.w.',
       '$448 p.w.'], dtype=object)

In [45]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('w\$')
].unique()

#yes, its recorded per week

array(['$425pw$1,847pcm'], dtype=object)

In [46]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('w/')
].unique()

#yes, its recorded per week

array(['$460pw/$1999pcm', '$450pw/$1955pcm', '$520pw/$2,260pcm',
       '$540pw/$2,346pcm', '$350pw/$1,521pcm', '$780pw/$3,390pcm',
       '$520pw/$2,259pcm', '$545pw/$2368pcm', '$675pw/$2933pcm',
       '$310pw/$1347pcm', '$610pw/$2651pm', '$580pw/$2530pcm',
       '$750pw/$3259pcm', '$350 pw/ $1521pm', '$550pw/$2390pcm',
       '$400pw/$1738pcm', '$580pw/$2520pcm', '$680pw/$2955pcm',
       '$850pw/$3693pcm', '$510pw/$2216pcm', '$440pw/$1912pcm',
       '$430pw/$1868pcm', '$485pw/$2107pcm', '$600pw/$2607pcm',
       '$875pw/$3802pcm', '$615pw/$2665pcm', '$420pw/$1825pcm',
       '$540pw/$2346pcm', '$650pw/ $2824 pcm', '$450pw/$1,955pcm',
       '$650pw/$2,824pcm', '$400pw/$1,738pcm', '$625pw/$2,716pcm',
       '$490pw/$2,129pcm', '$410pw/$1,782pcm', '$500pw/$2,172pcm',
       '$950pw/$4,128pcm', '$375pw/$1,629pcm', '$390pw/$1695pcm',
       '$400pw/$1,739pcm', '$700pw/$3042pcm', '$640pw/$2781pcm',
       '$800pw/$3476pcm', '$480pw/$2086pcm', '$795pw/$3454pcm',
       '$475pw/$2,064pc

In [47]:
rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index][
    rental_raw_data.cost_text.loc[test_rental_raw_cost_contains_w.index].\
        str.contains('w/\$')
].unique()

#yes, its recorded per week

array(['$460pw/$1999pcm', '$450pw/$1955pcm', '$520pw/$2,260pcm',
       '$540pw/$2,346pcm', '$350pw/$1,521pcm', '$780pw/$3,390pcm',
       '$520pw/$2,259pcm', '$545pw/$2368pcm', '$675pw/$2933pcm',
       '$310pw/$1347pcm', '$610pw/$2651pm', '$580pw/$2530pcm',
       '$750pw/$3259pcm', '$550pw/$2390pcm', '$400pw/$1738pcm',
       '$580pw/$2520pcm', '$680pw/$2955pcm', '$850pw/$3693pcm',
       '$510pw/$2216pcm', '$440pw/$1912pcm', '$430pw/$1868pcm',
       '$485pw/$2107pcm', '$600pw/$2607pcm', '$875pw/$3802pcm',
       '$615pw/$2665pcm', '$420pw/$1825pcm', '$540pw/$2346pcm',
       '$450pw/$1,955pcm', '$650pw/$2,824pcm', '$400pw/$1,738pcm',
       '$625pw/$2,716pcm', '$490pw/$2,129pcm', '$410pw/$1,782pcm',
       '$500pw/$2,172pcm', '$950pw/$4,128pcm', '$375pw/$1,629pcm',
       '$390pw/$1695pcm', '$400pw/$1,739pcm', '$700pw/$3042pcm',
       '$640pw/$2781pcm', '$800pw/$3476pcm', '$480pw/$2086pcm',
       '$795pw/$3454pcm', '$475pw/$2,064pcm', '$370pw/$1608pcm',
       '$820pw/$3,563pcm'

Now.. look for instances for which <code>re.findall</code> returns <code>'w'</code> only

In [48]:
cost_matches_to_w_only = test_rental_raw_cost_contains_w.map(lambda x: x[0])[
    test_rental_raw_cost_contains_w.map(lambda x: x[0]) == 'w'
]

In [49]:
sorted(rental_raw_data['cost_text'].loc[cost_matches_to_w_only.index].\
    str.findall(r'.{0,1}w').\
    str.join(', ').\
    unique()
    )

[' w',
 ' w, ew',
 ' w, ow',
 ' w, pw',
 '.w',
 '/w',
 '/w, ow',
 'ow, /w',
 'ow, /w, /w',
 'pw',
 'pw, ew',
 'pw, ow',
 'pw, pw',
 'pw, pw, ow']

It's obvious that <code>cost_text</code>s containing <code>pw</code> signify weekly rent. But what about <code>ew, ow</code>?

In [50]:
rental_raw_data['cost_text'][
    rental_raw_data['cost_text'].str.contains('ew')
]

659      leased - more properties wanted call andrew 04...
754           $450 pw - save time view our 3d tour & apply
1712          $650 pw - save time view our 3d tour & apply
1867                            $510 per week - new carpet
4315          $500 pw - save time view our 3d tour & apply
5225                       large 2bedroom with river views
5265       $650/w, brand new unit with stunning city views
7133                  $425pw - s/cage! white goods! views!
9548          $600 pw - save time view our 3d tour & apply
9815                     brand new & only $400.00 per week
9863                        $485 pw reviewing applications
9987                        $420 per week - new floorboard
10389                       $620pw - study! s/cage! views!
10716        $525pw - art precinct views plus storage cage
10759     $580pw - great views and nbn connection paid for
11771         $900 pw - save time view our 3d tour & apply
13998                    $750 **level 53 high level view

In [51]:
rental_raw_data['cost_text'][
    rental_raw_data['cost_text'].str.contains('ow')
].unique()

array(['enquire now: $300 per week', 'no deposit required or rent2own',
       '$580/w & available now', '$450 pw / avail. now',
       '$375 pw / avail. now', '$430 pw / avail. now',
       'rent2own with no deposit', '$460/w & available now',
       'enquire now: $490 per week', 'enquire now: $400 per week',
       'rent2own program', '$1000 pw high, $650 pw low, $160 pn',
       '$1200 pw high, $1000 pw low',
       '$1750 pw high, $1400 pw low, $250 pn',
       '$1500 pw high, $1050 pw low',
       '$2,000 pw high, $1,200 pw low, $300 pn',
       '$1200 pw high, $600 pw low, $180 pn',
       '$1750 pw high, $1400 pw low',
       '$2500 pw high, $1600 pw low, $380 pn',
       '$1,200 pw high, $950 pw low, $190 pn',
       '$1200 pw high, $750 pw low, $145 pn',
       'no bond required with rent2own',
       '$330 per week!! available now !!', 'rent2own',
       '$420 per week!! enquire now!!', 'available now from $450 p/w',
       'available now. from $380 p/w', 'available now from 

Note that there are some instances that do not actually have rental prices, but our regex for storing the matches in <code>test_rental_raw_cost_contains_w</code> falsely interpreted the numbers as rental prices.

- For instance, one value of <code>cost_text</code> is <code>'rent2own with no deposit'</code>. Here <code>2</code> is a number and there's a <code>w</code> following it after a few characters later.

A starting point is to filter out instances that have numbers for which digit is less than 2 or bigger than 3. 



I should've done this earlier, but partition <code>rental_raw_data['cost_text']</code> to the ones matching with instances in <code>test_rental_raw_cost_contains_w</code>

In [52]:
rental_data_cost_text_contains_w = rental_raw_data[
            rental_raw_data.cost_text.index.isin(test_rental_raw_cost_contains_w.index)
            ]['cost_text'].copy()

Inspect instances with values <code>cost_text</code> similar to <code>'rent2own with no deposit'</code>

In [53]:
rental_data_cost_text_contains_w[
    rental_data_cost_text_contains_w.str.findall(r'^[^\d]+\d[^\d]+$').map(lambda x: bool(x))
].unique()

array(['no deposit required or rent2own', 'rent2own with no deposit',
       'rent2own program', 'leased in 1 week!',
       'no bond required with rent2own',
       'large 2bedroom with river views', 'rent2own'], dtype=object)

In [54]:
rental_data_cost_text_no_costs_idx = rental_data_cost_text_contains_w[
    rental_data_cost_text_contains_w.str.findall(r'^[^\d]+\d[^\d]+$').map(lambda x: bool(x))
].index

<font color='blue'>Remove instances that don't have actual prices listed on <code>cost_text</code></font>
- Blue fonts: make changes in another variable but not in the actual dataset

In [55]:
rental_data_cost_text_contains_w = rental_data_cost_text_contains_w[~rental_data_cost_text_contains_w.index.isin(rental_data_cost_text_no_costs_idx)]

In [56]:
rental_data_cost_text_contains_w[
    rental_data_cost_text_contains_w.str.findall(r'\d{4,5}').map(lambda x: bool(x))
]

73                                      $1300 week
77               $545 per week week until jan 2023
125                                $460pw/$1999pcm
187                                     4125000 pw
325                            $300 pw / $1304 pcm
                           ...                    
14552    $1600 per week including pool maintenance
14580                              $675pw/$2933pcm
14581                              $600pw/$2607pcm
14627                              $530pw/$2303pcm
14801                       $440p/w ($1911.90 pcm)
Name: cost_text, Length: 721, dtype: object

<font color='blue'>What happens if we try to DIRECTLY extract the numbers from <code>rental_data_cost_text_contains_w</code>?</font>

In [57]:
rental_data_cost_text_contains_w

5                 $550 per week
10                $430 per week
27                $370 per week
29                $1,150 weekly
30                  $590 weekly
                  ...          
14790             $950 per week
14791             $460 per week
14794             $640 per week
14796               $525 weekly
14801    $440p/w ($1911.90 pcm)
Name: cost_text, Length: 8826, dtype: object

In [58]:
rental_data_cost_text_contains_w.str.findall(r'(\d+).*w[^\s\d]*')

5        [550]
10       [430]
27       [370]
29         [1]
30       [590]
         ...  
14790    [950]
14791    [460]
14794    [640]
14796    [525]
14801    [440]
Name: cost_text, Length: 8826, dtype: object

In [59]:
rental_data_cost_text_contains_w.loc[29]

'$1,150 weekly'

<font color='red'>Problem: looking at the 4th match, the syntax above wouldn't be able to pick up prices that have commas, e.g. <code>$1,150</code> </font>

A solution: find all such instances, then replace all ',' with ''.

In [60]:
cost_text_contains_comma_thousand_separator = rental_data_cost_text_contains_w[
    rental_data_cost_text_contains_w.str.findall(r'\d+,\d+').map(lambda x: bool(x))
]

<font color='blue'>Replace the commas in all instances whose prices are separated by commas with empty string</font>

In [61]:
rental_data_cost_text_contains_w.loc[cost_text_contains_comma_thousand_separator.index] = rental_data_cost_text_contains_w.str.replace(',', '')

In [62]:
rental_data_cost_text_contains_w.loc[cost_text_contains_comma_thousand_separator.index]

29                    $1150 weekly
101      $560 per week / $2433 pcm
327              $420pw / $1825pcm
435              $405pw | $1760pcm
437              $400pw | $1738pcm
                   ...            
14517              $750pw/$3259pcm
14519              $465pw/$2020pcm
14557               $1190 per week
14563               $1300 per week
14607              $600pw/$2607pcm
Name: cost_text, Length: 323, dtype: object

In [63]:
rental_data_cost_text_contains_w.str.findall(r'\d+.*\d+')

5                     [550]
10                    [430]
27                    [370]
29                   [1150]
30                    [590]
                ...        
14790                 [950]
14791                 [460]
14794                 [640]
14796                 [525]
14801    [440p/w ($1911.90]
Name: cost_text, Length: 8826, dtype: object

Now that's better.

We can re-extract the rental prices again: 

In [64]:
rental_data_cost_text_contains_w_mapped_prices = rental_data_cost_text_contains_w.str.findall(r'(\d+).*w[^\s\d]*').map(lambda x: x[0]).astype('float').sort_values()

In [65]:
rental_data_cost_text_contains_w_mapped_prices

12428          1.0
13537          5.0
604           26.0
600           28.0
2961          33.0
           ...    
2830        3775.0
13642       4500.0
13356       5000.0
12788       5500.0
187      4125000.0
Name: cost_text, Length: 8826, dtype: float64

In [66]:
#prices seem fishy for first 2 instances
rental_data_cost_text_contains_w.loc[[12428, 13537]]

12428    1 weeks free rent then $165 per week all inclu...
13537        5 mins to tram & shops & school,$880/per week
Name: cost_text, dtype: object

In [67]:
#prices seem fishy for the last instance (what? 4125000 per WEEK?)
rental_data_cost_text_contains_w.loc[187]

'4125000 pw'

In [68]:
rental_raw_data.loc[187,:]

Unnamed: 0     https://www.domain.com.au/level-3-302-13-15-la...
coordinates                           [-37.7316459, 144.7446886]
cost_text                                             4125000 pw
desc           - Centrally located with tranquil views.\n- Le...
name           Level 3, 302/13-15 Lake Street Caroline Spring...
rooms                                                         []
Name: 187, dtype: object

We'll just keep this "outlier" for further prerocessing.

Also, it seems that our regex fails to pick up the correct rental prices for both instances above (should be 165 an 880 respectively, not 1 and 5)

<font color='blue'>We can manually hardcode the prices for those 2 instances, it's just 2 instances anyway </font>

In [69]:
rental_data_cost_text_contains_w_mapped_prices.loc[[12428, 13537]]  = [165,880]

In [70]:
rental_raw_data

Unnamed: 0  \
0      https://www.domain.com.au/15-fintonia-street-h...   
1      https://www.domain.com.au/41-bishop-street-oak...   
2      https://www.domain.com.au/109-6-dalgety-street...   
3      https://www.domain.com.au/16-30-swindon-road-o...   
4      https://www.domain.com.au/1-14-bletchley-road-...   
...                                                  ...   
14799  https://www.domain.com.au/8a-logan-street-mans...   
14800  https://www.domain.com.au/27-elvins-street-man...   
14801  https://www.domain.com.au/1-30-ailsa-street-ma...   
14802  https://www.domain.com.au/49-hunter-street-man...   
14803  https://www.domain.com.au/6-clarke-street-mans...   

                             coordinates               cost_text  \
0             [-37.9032708, 145.0770553]                    $680   
1              [-37.9077745, 145.092738]                 $650.00   
2             [-37.8922004, 145.0899378]                    $500   
3             [-37.8972564, 145.0871089]                    $570   
4               [-37.894307, 145.082706]                    $475   
...                                  ...                     ...   
14799         [-37.0476758, 146.0845677]                    $400   
14800  [-37.05395439999999, 146.0797784]                    $480   
14801         [-37.0551267, 146.0929647]  $440p/w ($1911.90 pcm)   
14802         [-37.0561829, 146.0891432]                    $430   
14803         [-37.0494926, 146.0876812]                    $380   

                                                    desc  \
0                                Joseph Chatziconstantis   
1                                      James Drakopoulos   
2                                             Nigel Chee   
3                                     Matthew Swinnerton   
4                                               Jason Gu   
...                                                  ...   
14799   class="css-dxogle">* Unverified feature<svg a...   
14800   class="css-dxogle">* Unverified feature<svg a...   
14801   class="css-dxogle">* Unverified feature<svg a...   
14802   class="css-dxogle">* Unverified feature<svg a...   
14803   class="css-dxogle">* Unverified feature<svg a...   

                                          name  \
0       15 Fintonia Street Hughesdale VIC 3166   
1           41 Bishop Street Oakleigh VIC 3166   
2       109/6 Dalgety Street Oakleigh VIC 3166   
3         16/30 Swindon Road Oakleigh VIC 3166   
4      1/14 Bletchley Road Hughesdale VIC 3166   
...                                        ...   
14799       8A Logan Street Mansfield VIC 3722   
14800      27 Elvins Street Mansfield VIC 3722   
14801     1/30 Ailsa Street Mansfield VIC 3722   
14802      49 Hunter Street Mansfield VIC 3722   
14803       6 Clarke Street Mansfield VIC 3722   

                                    rooms  
0      ['3 Beds', '2 Baths', '2 Parking']  
1      ['3 Beds', '2 Baths', '2 Parking']  
2        ['1 Bed', '1 Bath', '1 Parking']  
3      ['3 Beds', '2 Baths', '2 Parking']  
4       ['2 Beds', '1 Bath', '1 Parking']  
...                                   ...  
14799   ['3 Beds', '1 Bath', '2 Parking']  
14800  ['3 Beds', '2 Baths', '4 Parking']  
14801   ['2 Beds', '1 Bath', '3 Parking']  
14802   ['3 Beds', '1 Bath', '3 Parking']  
14803   ['2 Beds', '1 Bath', '2 Parking']  

[14472 rows x 6 columns]

# Make sure that the  rest of the instances (other than the ones containing <code>w</code>) are NOT weekly rents

In [71]:
cost_text_not_containing_w = rental_raw_data.cost_text[~rental_raw_data.cost_text.index.isin(test_rental_raw_cost_contains_w.index)]

cost_text_not_containing_w[
    cost_text_not_containing_w.str.contains('w')
]

659    leased - more properties wanted call andrew 04...
Name: cost_text, dtype: object

Yeah, it seems that its one of the instances that don't actually have the rental price listed but included *just because* there's numeric digits

Now, these are the remaning instances for which <code>cost_text</code> should be processed:

In [72]:
cost_text_not_containing_w.sort_values(ascending=False)

13155                                    unfurnished $530
262                                under application $480
11323    rooming house $880 per month all bills inclusive
603                         prices starting from $155 pcm
11903                           peak $2500/off peak $1500
                               ...                       
1589                                               $1,000
13465                                              $1,000
7135                                                $ 600
11018                                               $ 160
7549                    $ 125 to $ 190 per calendar month
Name: cost_text, Length: 5600, dtype: object

Remove comma thousand separators. 
but first, remove such instances where theres ',' for numeric separators

In [73]:
cost_text_no_w_contains_comma_thousand_separator = cost_text_not_containing_w[
    cost_text_not_containing_w.str.findall(r'\d+,\d+').map(lambda x: bool(x))
]

In [74]:
cost_text_no_w_contains_comma_thousand_separator.unique()

array(['$1,200.00', '$1,000', '$1,199', '$1,400', '$2,800', '$1,280',
       '$1,100.00', '$1,150.00', '$1,295.00', '$1,000.00', '$2,500.00',
       '$39,000 p.a + outgoings',
       '$3,000 per month, plus gst & outgoings', '$1,650.00', '$1,250.00',
       '$1,080.00', '$1,500.00', '$1,350', '$1,200', '$1,500',
       '$1,025.00', '$1,100', '$1,050.00', '$1,350.00', '$1,300.00',
       '$1,450.00', '$1,550.00', '$1,125', '$1,600.00', '$1,495.00',
       '$1,150', '$1,325.00', '$1,495', '($1,350 pcm + $2,150 pcm)',
       '$2,000', '$1,690 pm', '$3,850', '$2,000.00', '$1,190', '$1,600',
       '$1,075.00', '$1,300', '$1,975', '$2,700', '$5,000', '$1,750.00',
       '$2,500', '$3,000', '$1,050', '$1,400.00', '$85,000.00',
       '$1,900 per month', '$1,700.00', '$1,799', '$1,500 mth',
       '$1,800.00', '$1,360.00', '$1,395.00', '$2,100.00',
       '$38,000 p.a. incl. outgoings + gst', '$1,907 calendar month',
       '$2,200 per month', '$47,000 per annum', '$2,200.00', '$1,615',
     

If the commas only occur to separate digits in rental prices then we'll just replace commas with '' for the entire <code>cost_text_not_containing_w</code>

Check if there's any other instances containing ','s but not serving as thousand separators

In [75]:
cost_text_not_containing_w[~cost_text_not_containing_w.index.isin(cost_text_no_w_contains_comma_thousand_separator.index)][
    cost_text_not_containing_w[~cost_text_not_containing_w.index.isin(cost_text_no_w_contains_comma_thousand_separator.index)].str.contains(',')
]

6666    $368 per month, $1600 per month
Name: cost_text, dtype: object

So we can replace all instances of ','s throughout all rows in <code>cost_text_not_containing_w</code>

In [76]:
cost_text_not_containing_w = cost_text_not_containing_w.str.replace(',', '')

# Make sure that the  rest of the instances (other than the ones containing <code>w</code>) are NOT weekly rents: start by finding all instances containing <code>p</code>
- Idea is to find prices that are of different granularity (e.g. per annum, etc)
- Even if it doesn't indicate any granularity for the prices, as usual we'll assume that they're weekly 

In [77]:
cost_text_not_containing_w[
    cost_text_not_containing_w.str.contains('\d+.*p')
]

167                                    $41250 pa
565                         from $1000 per month
603                prices starting from $155 pcm
622                          from $260 per night
623                               $220 per night
                          ...                   
13550    $120000 plus gst and plus outgoings pa.
13699                             $280 per month
13703                             $270 per month
13982           $550 **inspection upon request**
14198                   $420.00pm / $ 1825.00pcm
Name: cost_text, Length: 76, dtype: object

In [78]:
syntax = r'(p.*[^\s]*)'

cost_text_not_containing_w_other_granularity = cost_text_not_containing_w.str.findall(syntax)[
    cost_text_not_containing_w.str.findall(syntax).map(lambda x: bool(x))
]


print("Max length for matches in each instance: ",cost_text_not_containing_w_other_granularity.map(lambda x: len(x)).max())

print(cost_text_not_containing_w_other_granularity.map(lambda x: x[0]).unique())


Max length for matches in each instance:  1
['pa' 'pplication $480' 'per month' 'prices starting from $155 pcm'
 'per night' 'per night*' 'properties wanted call andrew 0407 060 344'
 'p.a + outgoings' 'per month plus gst & outgoings' 'pcm'
 'per calendar month' 'per hour' 'ps/$1738pcm' 'pcm + $2150 pcm)' 'pm'
 'per night off peak' 'pacious large 3 bedrooms apartment**'
 'park facing 3-storey home**' 'p/m' 'per month $1600 per month'
 'pplication)' 'pcm / $2455' 'payments requested'
 'p.a. incl. outgoings + gst' 'per cal month'
 'price $450 more needed urgently' 'per annum' 'pection upon request**'
 'private inspection available**' 'partment in southbank**'
 'per month all bills inclusive' 'peak $2500/off peak $1500' 'pm/$1695pcm'
 'per night - 2 people' 'per night - 4 people' 'plus gst plus out goings.'
 'plus gst and plus outgoings pa.' 'pm / $ 1825.00pcm']


Assume that prices that have granularities specified follow the format in the regex below (a number followed by any character followed by *p*, as in e.g. *p*er annum, *p*er month, etc)

In [79]:
syntax = r'\d+.*(p.*[^\s]*)'

cost_text_not_containing_w_other_granularity = cost_text_not_containing_w.str.findall(syntax)[
    cost_text_not_containing_w.str.findall(syntax).map(lambda x: bool(x))
]

print("Max length for matches in each instance: ",cost_text_not_containing_w_other_granularity.map(lambda x: len(x)).max())

print(cost_text_not_containing_w_other_granularity.map(lambda x: x[0]).unique())

cost_text_no_w_but_contain_p = cost_text_not_containing_w.loc[
    cost_text_not_containing_w_other_granularity.map(lambda x: x[0]).index
]

Max length for matches in each instance:  1
['pa' 'per month' 'pcm' 'per night' 'per night*' 'p.a + outgoings'
 'plus gst & outgoings' 'per calendar month' 'per hour' 'pcm)' 'pm' 'peak'
 'partment**' 'park facing 3-storey home**' 'p/m' 'plication)'
 'pcm / $2455' 'payments requested' 'p.a. incl. outgoings + gst'
 'per cal month' 'per annum' 'pon request**' 'pection available**'
 'partment in southbank**' 'per month all bills inclusive' 'peak $1500'
 'ple' 'plus out goings.' 'pa.']


The unique phrases/words that start with <code>p</code> are not that many

Each instance has exactly one of these. So to convert to per week (for granularities e.g. <code>pa, per month, pcm</code> etc) we just need to collect all indicators representing each granularity and later convert them

In [80]:
#indicators for prices that havve their granularities specified
per_annum_indicators = ['pa', 'p.a + outgoings', 'p.a. incl. outgoings + gst', 'per annum', 'pa.']
per_month_indicators = ['per month', 'pcm', 'per calendar month', 'pcm)', 'pm', 'p/m', 'pcm / $2455', 'per cal month', 'per month all bills inclusive']
per_night_indicators = ['per night', 'per night*']

In [81]:
cost_text_not_containing_w_other_granularity.map(lambda x: x[0])[
    cost_text_not_containing_w_other_granularity.map(lambda x: x[0]).isin(per_annum_indicators)
]

167                              pa
1050                p.a + outgoings
9171     p.a. incl. outgoings + gst
10481                     per annum
13550                           pa.
Name: cost_text, dtype: object

In [82]:
per_annum_indicators_idx = cost_text_not_containing_w_other_granularity.map(lambda x: x[0])[
    cost_text_not_containing_w_other_granularity.map(lambda x: x[0]).isin(per_annum_indicators)
].index 

per_month_indicators_idx = cost_text_not_containing_w_other_granularity.map(lambda x: x[0])[
    cost_text_not_containing_w_other_granularity.map(lambda x: x[0]).isin(per_month_indicators)
].index 

per_night_indicators_idx = cost_text_not_containing_w_other_granularity.map(lambda x: x[0])[
    cost_text_not_containing_w_other_granularity.map(lambda x: x[0]).isin(per_night_indicators)
].index 

Convert rental prices with known granularities

In [83]:
NUM_DAYS_PER_NIGHT = 7
NUM_WEEKS_PER_MONTH_APPROX = 4
NUM_WEEKS_PER_ANNUM_APPROX = 52

cost_text_no_w_but_contain_p.loc[per_night_indicators_idx] = cost_text_no_w_but_contain_p.loc[per_night_indicators_idx].\
                                                    str.findall(r'(\d+).*p.*[^\s]*').\
                                                    map(lambda x: x[0]).astype('float') * NUM_DAYS_PER_NIGHT


cost_text_no_w_but_contain_p.loc[per_month_indicators_idx] = cost_text_no_w_but_contain_p.loc[per_month_indicators_idx].\
                                                    str.findall(r'(\d+).*p.*[^\s]*').\
                                                    map(lambda x: x[0]).astype('float') / NUM_WEEKS_PER_MONTH_APPROX

cost_text_no_w_but_contain_p.loc[per_annum_indicators_idx] = cost_text_no_w_but_contain_p.loc[per_annum_indicators_idx].\
                                                    str.findall(r'(\d+).*p.*[^\s]*').\
                                                    map(lambda x: x[0]).astype('float') / NUM_WEEKS_PER_ANNUM_APPROX                                                 

                

What about the rest of the instances that contain <code>p</code> but don't have the granularities clearly specified

In [84]:
no_granularity_but_contain_p_idx = list(set(cost_text_no_w_but_contain_p.index) - (set(per_night_indicators_idx).union(set(per_month_indicators_idx)).union(set(per_annum_indicators_idx))))

In [85]:
cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx]

2829                                     $50 per hour
1297             $3000 per month plus gst & outgoings
5653        $850 **unique park facing 3-storey home**
6814                         $325 (under application)
13982                $550 **inspection upon request**
5293     $850 **spacious large 3 bedrooms apartment**
12860                       $220 per night - 2 people
12861                       $250 per night - 4 people
10818                $550 **inspection upon request**
10819           $400 **private inspection available**
10964                $650 **inspection upon request**
5206                     from $400 per night off peak
5210                     from $300 per night off peak
5211                     from $180 per night off peak
5212                     from $320 per night off peak
7773              $800 fortnightly payments requested
5213                     from $280 per night off peak
10857       $600 **furnished apartment in southbank**
13430                $39000 

It seems that our regex has false negatives (there's instances that still have granularities: e.g. per night, per hour, etc)

In [86]:
hourly_indicators = ['per hour']
monthly_indicators = ['per month']
nightly_indicators = ['per night']
fornightly_indicators = ['fortnightly']
granularity_indicators = hourly_indicators + monthly_indicators + nightly_indicators + fornightly_indicators

In [87]:
cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx][
    (cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('per hour')) | 
    (cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('per month')) | 
    (cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('per night')) |
    (cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('fortnightly')) 
]

2829                             $50 per hour
1297     $3000 per month plus gst & outgoings
12860               $220 per night - 2 people
12861               $250 per night - 4 people
5206             from $400 per night off peak
5210             from $300 per night off peak
5211             from $180 per night off peak
5212             from $320 per night off peak
7773      $800 fortnightly payments requested
5213             from $280 per night off peak
Name: cost_text, dtype: object

In [88]:
cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx][
    (cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('per hour')) | 
    (cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('per month')) | 
    (cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('per night')) |
    (cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('fortnightly')) 
]

2829                             $50 per hour
1297     $3000 per month plus gst & outgoings
12860               $220 per night - 2 people
12861               $250 per night - 4 people
5206             from $400 per night off peak
5210             from $300 per night off peak
5211             from $180 per night off peak
5212             from $320 per night off peak
7773      $800 fortnightly payments requested
5213             from $280 per night off peak
Name: cost_text, dtype: object

In [89]:
cost_text_no_w_but_contain_p = cost_text_no_w_but_contain_p.astype('str')

Apparently to mutate changes only able to call <code>.loc</code> once

In [90]:
NUM_HOURS_PER_WEEK = 168

idx_to_be_updated = cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx][
    cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('per hour')
].index


cost_text_no_w_but_contain_p.loc[idx_to_be_updated]= \
    cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx][
    cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('per hour')
].str.findall(r'(\d+).*p').map(lambda x: x[0]).astype('float') * NUM_HOURS_PER_WEEK

cost_text_no_w_but_contain_p = cost_text_no_w_but_contain_p.astype('str')

In [91]:
idx_to_be_updated = cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx][
    cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('per night')
].index

cost_text_no_w_but_contain_p.loc[idx_to_be_updated] =\
    cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx][
    cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('per night')
].str.findall(r'(\d+).*p').map(lambda x: x[0]).astype('float') * NUM_DAYS_PER_NIGHT

cost_text_no_w_but_contain_p = cost_text_no_w_but_contain_p.astype('str')

In [92]:
idx_to_be_updated = cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx][
    cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('per month')
].index

cost_text_no_w_but_contain_p.loc[idx_to_be_updated] =\
    cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx][
    cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('per month')
].str.findall(r'(\d+).*p').map(lambda x: x[0]).astype('float') / NUM_WEEKS_PER_MONTH_APPROX

cost_text_no_w_but_contain_p = cost_text_no_w_but_contain_p.astype('str')

In [93]:
NUM_WEEKS_PER_FORTNIGHT = 2
idx_to_be_updated = cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx][
    cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('fortnightly')
].index

cost_text_no_w_but_contain_p.loc[idx_to_be_updated] =\
    cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx][
    cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('fortnightly')
].str.findall(r'(\d+).*p').map(lambda x: x[0]).astype('float') / NUM_WEEKS_PER_FORTNIGHT

cost_text_no_w_but_contain_p = cost_text_no_w_but_contain_p.astype('str')

Now, as for the rest of the instances that don't have any granularities but still contain <code>p</code>

In [94]:
undiscovered_granularities_contain_p = cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx][
    (cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('per hour')) | 
    (cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('per month')) | 
    (cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('per night')) |
    (cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].str.contains('fortnightly')) 
]

In [95]:
no_granularites_but_contain_p_idx = list(set(cost_text_no_w_but_contain_p.loc[no_granularity_but_contain_p_idx].index) -\
    set(undiscovered_granularities_contain_p.index))

In [96]:
#extract the numbers, convert them to numeric
cost_text_no_w_but_contain_p.loc[no_granularites_but_contain_p_idx] = cost_text_no_w_but_contain_p.loc[no_granularites_but_contain_p_idx].\
                                                                str.findall(r'(\d+).*p*').map(lambda x: x[0]).astype('float') 

In [97]:
cost_text_no_w_but_contain_p.astype('float')

167       793.269231
565       250.000000
603        38.750000
622      1820.000000
623      1540.000000
            ...     
13550    2307.692308
13699      70.000000
13703      67.500000
13982     550.000000
14198     105.000000
Name: cost_text, Length: 76, dtype: float64

In [98]:
cost_text_no_w_but_contain_p = cost_text_no_w_but_contain_p.astype('float')

In [99]:
cost_text_no_w_but_contain_p

167       793.269231
565       250.000000
603        38.750000
622      1820.000000
623      1540.000000
            ...     
13550    2307.692308
13699      70.000000
13703      67.500000
13982     550.000000
14198     105.000000
Name: cost_text, Length: 76, dtype: float64

# Now.. Inspect the instances that don't contain <code>p</code>

In [100]:
cost_text_not_containing_w[~cost_text_not_containing_w.index.isin(cost_text_no_w_but_contain_p.index)]

0           $680
1        $650.00
2           $500
3           $570
4           $475
          ...   
14795       $750
14799       $400
14800       $480
14802       $430
14803       $380
Name: cost_text, Length: 5524, dtype: object

In [101]:
cost_text_no_w_not_contain_p = cost_text_not_containing_w[~cost_text_not_containing_w.index.isin(cost_text_no_w_but_contain_p.index)]

We can find those that have alphabets in them

It turns out that monthly rents don't necessarily have to include smth like <code>per month</code>, can just have e.g. <code>monthly</code>, which is a challenge.

But at the moment we don't need to process them first. Keep them first (using the indexes), process them later

In [102]:
cost_text_no_w_not_contain_p[
    cost_text_no_w_not_contain_p.str.contains('month')
]

6441                       $250 monthly
9479               $2297 calendar month
9480               $2557 calendar month
9482               $2470 calendar month
9483               $1365 calendar month
9484               $1950 calendar month
9485               $2384 calendar month
9486               $1907 calendar month
9487               $2384 calendar month
9488               $1950 calendar month
9489               $2384 calendar month
9491               $2687 calendar month
11562                       $2500/month
13146    $500 **3-6 months lease term**
13893                        $250/month
Name: cost_text, dtype: object

In [103]:
cost_text_no_w_not_contain_p_months_idx = cost_text_no_w_not_contain_p[
    cost_text_no_w_not_contain_p.str.contains('month')
].index

In [104]:
cost_text_no_w_not_contain_p_months_idx

Int64Index([ 6441,  9479,  9480,  9482,  9483,  9484,  9485,  9486,  9487,
             9488,  9489,  9491, 11562, 13146, 13893],
           dtype='int64')

Now that we know there's months, we can see the instances for which <code>cost_text</code> contains alphabets but don't contain <code>month</code>

In [105]:
cost_text_no_w_not_contain_p[
    (cost_text_no_w_not_contain_p.str.contains(r'[a-z]')) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_months_idx))
].unique()[0:40]

array(['under application $480', '$600 inc. bills',
       'leased - more properties wanted call andrew 0407 060 344',
       'from $195', '$420 to $450', '$1000.00 - fully furnished',
       '$800 to $880', '$750 to $775', 'furnished $480', '$33 to $65',
       'from $280', '450 furnished', '$490 fully furnished',
       '$180.00 to $220.00', '$370 - fully furnished',
       '$620 - fully furnished', '$430 - fully furnished',
       '$580 fully furnished', '$620 fully furnished',
       '$390 - fully furnished', '$650 - fully furnished',
       '$500 - fully furnished', '$420 - fully furnished',
       '$680 - fully furnished', '$770 - fully furnished',
       '$640 - fully furnished', '$720 fully furnished',
       '$530 - fully furnished', '$790 - fully furnished',
       '$490.00 - fully furnished', '$630 - fully furnished',
       '$580 - fully furnished', '$590 fully furnished',
       '$550.00 - fully furnished', '$530 fully furnished',
       '$560 - fully furnished', '$430 ful

In [106]:
cost_text_no_w_not_contain_p[
    (cost_text_no_w_not_contain_p.str.contains(r'[a-z]')) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_months_idx))
].unique()[41:]

array(['$410 - fully furnished', '$800 - fully furnished',
       '$600 - fully furnished', '$570 - fully furnished',
       '$780 - fully furnished', '$460 - fully furnished',
       '$700 - fully furnished', '$260 to $280', '$170 to $240',
       '$1500 mth', '$38000+gst+outgoing', '$900 inc bills & furniture',
       'leased at record price $450 more needed urgently',
       'leased by century21!!', '$500.00 fully furnished',
       '$580.00 | leased', '$620 **fully furnished**', '$680 (neg)',
       '$35000 annually', '$580.00 fully furnished',
       '$480 - fully furnished', 'from $160',
       '$300 all bills included in rent',
       '$190 including all utility bills', 'unfurnished $530',
       '$650 fully furnished', '1bed + 1study $560', '$650 furnished'],
      dtype=object)

It seems that <code>furnish</code> keeps popping up and quite common, we can exclude that and see that's left.
- Furthermore, such instances typically are only of the form <code> number + furnished </code> or vice versa. 
    - This means that later we can simply extract any occurence of any number in instances with <code>furnish</code> in it

In [107]:
cost_text_no_w_not_contain_p_furnish_idx = cost_text_no_w_not_contain_p[
    cost_text_no_w_not_contain_p.str.contains('furnish')
].index

Now that we know there's months, we can see the instances for which <code>cost_text</code> contains alphabets but don't contain <code>month, furnish</code>

In [108]:
cost_text_no_w_not_contain_p[
    (cost_text_no_w_not_contain_p.str.contains(r'[a-z]')) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_months_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_furnish_idx))
].unique()

array(['under application $480', '$600 inc. bills',
       'leased - more properties wanted call andrew 0407 060 344',
       'from $195', '$420 to $450', '$800 to $880', '$750 to $775',
       '$33 to $65', 'from $280', '$180.00 to $220.00', '$260 to $280',
       '$170 to $240', '$1500 mth', '$38000+gst+outgoing',
       '$900 inc bills & furniture',
       'leased at record price $450 more needed urgently',
       'leased by century21!!', '$580.00 | leased', '$680 (neg)',
       '$35000 annually', 'from $160', '$300 all bills included in rent',
       '$190 including all utility bills', '1bed + 1study $560'],
      dtype=object)

There's 2 instances that contain <code>mth</code> and <code>annually</code>.

In [109]:
cost_text_no_w_not_contain_p_mth_idx = cost_text_no_w_not_contain_p[
    (cost_text_no_w_not_contain_p.str.contains(r'[a-z]')) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_months_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_furnish_idx)) & 
    (cost_text_no_w_not_contain_p.str.contains('mth'))
    ].index

cost_text_no_w_not_contain_p_annually_idx = cost_text_no_w_not_contain_p[
    (cost_text_no_w_not_contain_p.str.contains(r'[a-z]')) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_months_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_furnish_idx)) & 
    (cost_text_no_w_not_contain_p.str.contains('annual'))
    ].index

Now that we know there's months, we can see the instances for which <code>cost_text</code> contains alphabets but don't contain <code>month, furnish, mth, annually</code>

In [110]:
cost_text_no_w_not_contain_p[
    (cost_text_no_w_not_contain_p.str.contains(r'[a-z]')) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_months_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_furnish_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_mth_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_annually_idx))
].unique()

array(['under application $480', '$600 inc. bills',
       'leased - more properties wanted call andrew 0407 060 344',
       'from $195', '$420 to $450', '$800 to $880', '$750 to $775',
       '$33 to $65', 'from $280', '$180.00 to $220.00', '$260 to $280',
       '$170 to $240', '$38000+gst+outgoing',
       '$900 inc bills & furniture',
       'leased at record price $450 more needed urgently',
       'leased by century21!!', '$580.00 | leased', '$680 (neg)',
       'from $160', '$300 all bills included in rent',
       '$190 including all utility bills', '1bed + 1study $560'],
      dtype=object)

There are 3 instances with no rental prices at all 

In [111]:
cost_text_no_w_not_contain_p.str.contains('leased by century21!!')

0        False
1        False
2        False
3        False
4        False
         ...  
14795    False
14799    False
14800    False
14802    False
14803    False
Name: cost_text, Length: 5524, dtype: bool

In [112]:
cost_text_no_w_not_contain_p_to_delete_idx = cost_text_no_w_not_contain_p[
    (cost_text_no_w_not_contain_p.str.contains('leased - more properties wanted call andrew 0407 060 344')) |
    (cost_text_no_w_not_contain_p.str.contains('leased by century21!!')) 
].index

Now that we know there's months, we can see the instances for which <code>cost_text</code> contains alphabets but don't contain <code>month, furnish, mth, annually</code>

In [113]:
cost_text_no_w_not_contain_p[
    (cost_text_no_w_not_contain_p.str.contains(r'[a-z]')) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_months_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_furnish_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_mth_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_annually_idx))& 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_to_delete_idx))
].unique()

array(['under application $480', '$600 inc. bills', 'from $195',
       '$420 to $450', '$800 to $880', '$750 to $775', '$33 to $65',
       'from $280', '$180.00 to $220.00', '$260 to $280', '$170 to $240',
       '$38000+gst+outgoing', '$900 inc bills & furniture',
       'leased at record price $450 more needed urgently',
       '$580.00 | leased', '$680 (neg)', 'from $160',
       '$300 all bills included in rent',
       '$190 including all utility bills', '1bed + 1study $560'],
      dtype=object)

There are instances that don't state the exact price, it gives price range instead
- To be honest I have no prior knowledge to suitably determine which to "pick" (e.g. in <code>$800 to $880</code> I dont know which to pick, 800 or 880)
- So I'll just return the average.

In [114]:
def is_price_exact(list_price):
    #since we're using re.findall on strings, convert them to floats first
    list_price = [float(price) for price in list_price]
    
    if len(list_price) > 0:
        return sum(list_price) / len(list_price)


In [115]:
cost_text_no_w_not_contain_p_use_average_idx = cost_text_no_w_not_contain_p[
    (cost_text_no_w_not_contain_p.str.contains(r'[a-z]')) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_months_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_furnish_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_mth_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_annually_idx))& 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_to_delete_idx))
].index

I will permanently include changes below once I've determined the rest of the instances

But for now, we'll just keep the syntax here first so i can come back to it

In [116]:
#note i include the $ sign because all unique values above have their prices preceded by a dollar sign.
cost_text_no_w_not_contain_p[
    (cost_text_no_w_not_contain_p.str.contains(r'[a-z]')) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_months_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_furnish_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_mth_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_annually_idx))& 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_to_delete_idx))
].str.findall(r'\$(\d+)').map(lambda x: is_price_exact(x))

262        480.0
564        600.0
2164       195.0
2939       435.0
4660       840.0
4663       840.0
4727       762.5
5417        49.0
5664       280.0
6714       200.0
8146       270.0
8519       205.0
9170     38000.0
10176      900.0
10228      450.0
10438      580.0
11773      680.0
12645      160.0
12691      300.0
12696      190.0
14098      560.0
Name: cost_text, dtype: float64

If the index below returns empty, meaning we've exhausted all possibilities of handling instances with alphabets

In [117]:
cost_text_no_w_not_contain_p[
    (cost_text_no_w_not_contain_p.str.contains(r'[a-z]')) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_months_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_furnish_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_mth_idx)) & 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_annually_idx))& 
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_to_delete_idx))&
    (~cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_use_average_idx))
].index

Int64Index([], dtype='int64')

Which means, the rest of the instances NOT containign alphabets SHOULD purely contain the numeric price values
- Meaning that we've CORRECTLY partitioned <code>cost_text_no_w_not_contain_p</code> into 6 partitions: 
1. ONes containing <code>months</code>
2. Ones containing <code>furnish</code>
3. Ones containing <code>mth</code>
4. Ones containing <code>annually</code>
5. Ones to be deleted because there's no rental price listed
6. Ones that contain a range of prices (and exact prices), for which I will take the average 

In [118]:
cost_text_no_w_not_contain_p_alphabetic_idx = set(cost_text_no_w_not_contain_p[
    (cost_text_no_w_not_contain_p.str.contains(r'[a-z]')) 
].index) 

cost_text_no_w_not_contain_p_alphabetic_partitioned_idx = set(cost_text_no_w_not_contain_p[
    (cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_months_idx)) | 
    (cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_furnish_idx)) |  
    (cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_mth_idx)) | 
    (cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_annually_idx)) | 
    (cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_to_delete_idx)) |
    (cost_text_no_w_not_contain_p.index.isin(cost_text_no_w_not_contain_p_use_average_idx))
].index) 

print(cost_text_no_w_not_contain_p_alphabetic_idx - cost_text_no_w_not_contain_p_alphabetic_partitioned_idx)
print(cost_text_no_w_not_contain_p_alphabetic_idx - cost_text_no_w_not_contain_p_alphabetic_partitioned_idx)


set()
set()


# Good. Now we just have to check the rest that are neatly written down as numeric values

Check if all rental prices are written as numerical values **preceded by \$**

In [119]:
cost_text_no_w_not_contain_p_no_alphas = cost_text_no_w_not_contain_p[
    (~cost_text_no_w_not_contain_p.str.contains(r'[a-z]'))
]

Find instances for which rental prices are directly preceded by <code>$</code>

In [120]:
cost_text_no_w_not_contain_p[
    (~cost_text_no_w_not_contain_p.str.contains(r'[a-z]')) & 
    (cost_text_no_w_not_contain_p.str.findall(r'\$\d+').map(lambda x: bool(x)))
]

0           $680
1        $650.00
2           $500
3           $570
4           $475
          ...   
14795       $750
14799       $400
14800       $480
14802       $430
14803       $380
Name: cost_text, Length: 5397, dtype: object

Find instances for which rental prices are NOT (or not directly) preceded by <code>$</code>

In [121]:
cost_text_no_w_not_contain_p[
    (~cost_text_no_w_not_contain_p.str.contains(r'[a-z]')) & 
    (~cost_text_no_w_not_contain_p.str.findall(r'\$\d+').map(lambda x: bool(x)))
]

160          360
190          220
975          300
1017      450.00
1048      430.00
1049      430.00
1885         490
3367         420
3414         410
3560      370.00
4313      780.00
5407         420
6325        1000
6590         370
7088         430
7135       $ 600
7162     1180.00
7290         300
7698         320
9912      410.00
10274     787.00
11018      $ 160
12135        350
13991        550
Name: cost_text, dtype: object

Ok let's handle them one by one.

In [122]:
cost_text_no_w_no_contain_p_no_alphas_with_dollarsign = cost_text_no_w_not_contain_p[
    (~cost_text_no_w_not_contain_p.str.contains(r'[a-z]')) & 
    (cost_text_no_w_not_contain_p.str.findall(r'\$\d+').map(lambda x: bool(x)))
]

cost_text_no_w_no_contain_p_no_alphas_no_dollarsign = cost_text_no_w_not_contain_p[
    (~cost_text_no_w_not_contain_p.str.contains(r'[a-z]')) & 
    (~cost_text_no_w_not_contain_p.str.findall(r'\$\d+').map(lambda x: bool(x)))
]

In [123]:
#check that we've correctly partitioned them
cost_text_no_w_no_contain_p_no_alphas_with_dollarsign.shape[0] + \
    cost_text_no_w_no_contain_p_no_alphas_no_dollarsign.shape[0] == cost_text_no_w_not_contain_p_no_alphas.shape[0]

True

Now.. for the prices with \$ signs

In [124]:
cost_text_no_w_no_contain_p_no_alphas_with_dollarsign.str.findall(r'\$(\d+)').sort_values()

2691            [1]
530      [100, 200]
1043         [1000]
14393        [1000]
14392        [1000]
            ...    
8995          [995]
8837          [995]
14188         [995]
2120          [995]
4091          [999]
Name: cost_text, Length: 5397, dtype: object

In [125]:
cost_text_no_w_no_contain_p_no_alphas_with_dollarsign.loc[530]

'$100-$200'

There's instances with 2 matches, again, indicating a price range instead of exact price. 
Again, we'll just average them. 

Now.. for the prices without \$ signs

In [126]:
cost_text_no_w_no_contain_p_no_alphas_no_dollarsign

160          360
190          220
975          300
1017      450.00
1048      430.00
1049      430.00
1885         490
3367         420
3414         410
3560      370.00
4313      780.00
5407         420
6325        1000
6590         370
7088         430
7135       $ 600
7162     1180.00
7290         300
7698         320
9912      410.00
10274     787.00
11018      $ 160
12135        350
13991        550
Name: cost_text, dtype: object

In [127]:
cost_text_no_w_no_contain_p_no_alphas_no_dollarsign.str.findall(r'\d+')

160           [360]
190           [220]
975           [300]
1017      [450, 00]
1048      [430, 00]
1049      [430, 00]
1885          [490]
3367          [420]
3414          [410]
3560      [370, 00]
4313      [780, 00]
5407          [420]
6325         [1000]
6590          [370]
7088          [430]
7135          [600]
7162     [1180, 00]
7290          [300]
7698          [320]
9912      [410, 00]
10274     [787, 00]
11018         [160]
12135         [350]
13991         [550]
Name: cost_text, dtype: object

# Begin to process <code>cost_text_no_w_not_contain_p</code>, now that we know all the cases to handle

1. Fix up instances containing <code>months</code>

In [128]:
cost_text_no_w_not_contain_p

0           $680
1        $650.00
2           $500
3           $570
4           $475
          ...   
14795       $750
14799       $400
14800       $480
14802       $430
14803       $380
Name: cost_text, Length: 5524, dtype: object

In [129]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_months_idx]

6441                       $250 monthly
9479               $2297 calendar month
9480               $2557 calendar month
9482               $2470 calendar month
9483               $1365 calendar month
9484               $1950 calendar month
9485               $2384 calendar month
9486               $1907 calendar month
9487               $2384 calendar month
9488               $1950 calendar month
9489               $2384 calendar month
9491               $2687 calendar month
11562                       $2500/month
13146    $500 **3-6 months lease term**
13893                        $250/month
Name: cost_text, dtype: object

It's sufficient to extract numbers after the dollar sign to extract the rental prices.

In [130]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_months_idx] = \
    cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_months_idx].str.findall(r'\$(\d+)').map(lambda x: x[0]).astype('float') / NUM_WEEKS_PER_MONTH_APPROX

#type cast into string first, in case it throws an error when we want to convert  all of them eventually into floats because ive ppotentially missed some cases
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_months_idx] = cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_months_idx].astype('str')

2. Fix up instances containing <code>furnish</code>

In [131]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_furnish_idx].unique()

array(['$1000.00 - fully furnished', 'furnished $480', '450 furnished',
       '$490 fully furnished', '$370 - fully furnished',
       '$620 - fully furnished', '$430 - fully furnished',
       '$580 fully furnished', '$620 fully furnished',
       '$390 - fully furnished', '$650 - fully furnished',
       '$500 - fully furnished', '$420 - fully furnished',
       '$680 - fully furnished', '$770 - fully furnished',
       '$640 - fully furnished', '$720 fully furnished',
       '$530 - fully furnished', '$790 - fully furnished',
       '$490.00 - fully furnished', '$630 - fully furnished',
       '$580 - fully furnished', '$590 fully furnished',
       '$550.00 - fully furnished', '$530 fully furnished',
       '$560 - fully furnished', '$430 fully furnished',
       '$540 fully furnished', '$490 - fully furnished',
       '$520 - fully furnished', '$670 - fully furnished',
       '$410 - fully furnished', '$800 - fully furnished',
       '$600 - fully furnished', '$570 - fully furnis

It's sufficient to extract all the numbers (not necessarily preceded by \$)

In [132]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_furnish_idx].str.findall(r'(\d+)').map(lambda x:x[0]).unique()

array(['1000', '480', '450', '490', '370', '620', '430', '580', '390',
       '650', '500', '420', '680', '770', '640', '720', '530', '790',
       '630', '590', '550', '560', '540', '520', '670', '410', '800',
       '600', '570', '780', '460', '700'], dtype=object)

In [133]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_furnish_idx] = \
    cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_furnish_idx].str.findall(r'(\d+)').map(lambda x: x[0])

#type cast into string first, in case it throws an error when we want to convert  all of them eventually into floats because ive ppotentially missed some cases
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_furnish_idx] = cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_furnish_idx].astype('str')

In [134]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_furnish_idx].unique()

array(['1000', '480', '450', '490', '370', '620', '430', '580', '390',
       '650', '500', '420', '680', '770', '640', '720', '530', '790',
       '630', '590', '550', '560', '540', '520', '670', '410', '800',
       '600', '570', '780', '460', '700'], dtype=object)

3. Fix up instances containing <code>mth</code>

In [135]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_mth_idx]

8765    $1500 mth
Name: cost_text, dtype: object

Since only one instance, no need to use regex

In [136]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_mth_idx] = str(1500 / NUM_WEEKS_PER_MONTH_APPROX)

In [137]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_mth_idx]

8765    375.0
Name: cost_text, dtype: object

4. Fix up instances containing <code>annually</codde>

In [138]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_annually_idx]

11823    $35000 annually
Name: cost_text, dtype: object

Since only one instance, no need to use regex

In [139]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_annually_idx] = str(35000 / NUM_WEEKS_PER_ANNUM_APPROX)

In [140]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_annually_idx]

11823    673.0769230769231
Name: cost_text, dtype: object

5. Delete instances that don't have rental prices listed

In [141]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_to_delete_idx]

659      leased - more properties wanted call andrew 04...
10233                                leased by century21!!
Name: cost_text, dtype: object

In [142]:
cost_text_no_w_not_contain_p = cost_text_no_w_not_contain_p.drop(cost_text_no_w_not_contain_p_to_delete_idx)

6. instances that have alphabets in it, and some of them have price range

In [143]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_use_average_idx] = \
    cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_use_average_idx].str.findall(r'\$(\d+)').map(lambda x: is_price_exact(x))

cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_use_average_idx] = cost_text_no_w_not_contain_p.loc[cost_text_no_w_not_contain_p_use_average_idx].astype('str')

7. instances that have no alphabets, with \$ sign

In [144]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_no_contain_p_no_alphas_with_dollarsign.index] =\
    cost_text_no_w_no_contain_p_no_alphas_with_dollarsign.str.findall(r'\$(\d+)').map(lambda x: is_price_exact(x))

cost_text_no_w_not_contain_p.loc[cost_text_no_w_no_contain_p_no_alphas_with_dollarsign.index] =\
    cost_text_no_w_not_contain_p.loc[cost_text_no_w_no_contain_p_no_alphas_with_dollarsign.index].astype('str')

8. instances that have no alphabets, without \$ sign

In [145]:
cost_text_no_w_not_contain_p.loc[cost_text_no_w_no_contain_p_no_alphas_no_dollarsign.index] =\
    cost_text_no_w_no_contain_p_no_alphas_no_dollarsign.str.findall(r'(\d+)').map(lambda x: is_price_exact(x))

cost_text_no_w_not_contain_p.loc[cost_text_no_w_no_contain_p_no_alphas_no_dollarsign.index] =\
    cost_text_no_w_not_contain_p.loc[cost_text_no_w_no_contain_p_no_alphas_no_dollarsign.index].astype('str')

In [146]:
cost_text_no_w_not_contain_p.astype('float')

0        680.0
1        650.0
2        500.0
3        570.0
4        475.0
         ...  
14795    750.0
14799    400.0
14800    480.0
14802    430.0
14803    380.0
Name: cost_text, Length: 5522, dtype: float64

# FINISHED

Here's the parsed <code>cost_text</code> for those instances containing <code>w</code>

In [147]:
rental_data_cost_text_contains_w_mapped_prices

12428        165.0
13537        880.0
604           26.0
600           28.0
2961          33.0
           ...    
2830        3775.0
13642       4500.0
13356       5000.0
12788       5500.0
187      4125000.0
Name: cost_text, Length: 8826, dtype: float64

Here's the parsed <code>cost_text</code> for those instances not containing <code>w</code> but containing <code>p</code> 

In [148]:
cost_text_no_w_but_contain_p

167       793.269231
565       250.000000
603        38.750000
622      1820.000000
623      1540.000000
            ...     
13550    2307.692308
13699      70.000000
13703      67.500000
13982     550.000000
14198     105.000000
Name: cost_text, Length: 76, dtype: float64

Here's the parsed <code>cost_text</code> for those instances not containing <code>w</code> but not containing <code>p</code> as well

In [149]:
cost_text_no_w_not_contain_p

0        680.0
1        650.0
2        500.0
3        570.0
4        475.0
         ...  
14795    750.0
14799    400.0
14800    480.0
14802    430.0
14803    380.0
Name: cost_text, Length: 5522, dtype: object

Most important thing: check if those 3 Series above have exhaustively covered all instances in the raw data!!

In [150]:
rental_raw_data

Unnamed: 0  \
0      https://www.domain.com.au/15-fintonia-street-h...   
1      https://www.domain.com.au/41-bishop-street-oak...   
2      https://www.domain.com.au/109-6-dalgety-street...   
3      https://www.domain.com.au/16-30-swindon-road-o...   
4      https://www.domain.com.au/1-14-bletchley-road-...   
...                                                  ...   
14799  https://www.domain.com.au/8a-logan-street-mans...   
14800  https://www.domain.com.au/27-elvins-street-man...   
14801  https://www.domain.com.au/1-30-ailsa-street-ma...   
14802  https://www.domain.com.au/49-hunter-street-man...   
14803  https://www.domain.com.au/6-clarke-street-mans...   

                             coordinates               cost_text  \
0             [-37.9032708, 145.0770553]                    $680   
1              [-37.9077745, 145.092738]                 $650.00   
2             [-37.8922004, 145.0899378]                    $500   
3             [-37.8972564, 145.0871089]                    $570   
4               [-37.894307, 145.082706]                    $475   
...                                  ...                     ...   
14799         [-37.0476758, 146.0845677]                    $400   
14800  [-37.05395439999999, 146.0797784]                    $480   
14801         [-37.0551267, 146.0929647]  $440p/w ($1911.90 pcm)   
14802         [-37.0561829, 146.0891432]                    $430   
14803         [-37.0494926, 146.0876812]                    $380   

                                                    desc  \
0                                Joseph Chatziconstantis   
1                                      James Drakopoulos   
2                                             Nigel Chee   
3                                     Matthew Swinnerton   
4                                               Jason Gu   
...                                                  ...   
14799   class="css-dxogle">* Unverified feature<svg a...   
14800   class="css-dxogle">* Unverified feature<svg a...   
14801   class="css-dxogle">* Unverified feature<svg a...   
14802   class="css-dxogle">* Unverified feature<svg a...   
14803   class="css-dxogle">* Unverified feature<svg a...   

                                          name  \
0       15 Fintonia Street Hughesdale VIC 3166   
1           41 Bishop Street Oakleigh VIC 3166   
2       109/6 Dalgety Street Oakleigh VIC 3166   
3         16/30 Swindon Road Oakleigh VIC 3166   
4      1/14 Bletchley Road Hughesdale VIC 3166   
...                                        ...   
14799       8A Logan Street Mansfield VIC 3722   
14800      27 Elvins Street Mansfield VIC 3722   
14801     1/30 Ailsa Street Mansfield VIC 3722   
14802      49 Hunter Street Mansfield VIC 3722   
14803       6 Clarke Street Mansfield VIC 3722   

                                    rooms  
0      ['3 Beds', '2 Baths', '2 Parking']  
1      ['3 Beds', '2 Baths', '2 Parking']  
2        ['1 Bed', '1 Bath', '1 Parking']  
3      ['3 Beds', '2 Baths', '2 Parking']  
4       ['2 Beds', '1 Bath', '1 Parking']  
...                                   ...  
14799   ['3 Beds', '1 Bath', '2 Parking']  
14800  ['3 Beds', '2 Baths', '4 Parking']  
14801   ['2 Beds', '1 Bath', '3 Parking']  
14802   ['3 Beds', '1 Bath', '3 Parking']  
14803   ['2 Beds', '1 Bath', '2 Parking']  

[14472 rows x 6 columns]

In [151]:
parsed_cost_text_idx = list(set(rental_data_cost_text_contains_w_mapped_prices.index).union(set(cost_text_no_w_but_contain_p.index)).union(set(cost_text_no_w_not_contain_p.index)))

In [152]:
len(parsed_cost_text_idx)

14424

We note that we deleted some instances that don't have the rental prices listed. We then verify that that is the case.

In [153]:
deleted_instances_idx = set(rental_raw_data.index) - set(parsed_cost_text_idx)

In [154]:
rental_raw_data.loc[deleted_instances_idx,:].cost_text.unique()

/tmp/ipykernel_27731/1437132011.py:1: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.
  rental_raw_data.loc[deleted_instances_idx,:].cost_text.unique()


array(['rent2own with no deposit', 'no deposit required or rent2own',
       'rent2own program',
       'leased - more properties wanted call andrew 0407 060 344',
       'no bond required with rent2own', 'leased by century21!!',
       'rent2own', 'leased in 1 week!', 'large 2bedroom with river views'],
      dtype=object)

# <font color='green'>Exclude the instances that don't have rental prices! </font>

In [155]:
rental_raw_data = rental_raw_data.loc[parsed_cost_text_idx,:]

Copy the dataframe first to test the change so that any unwanted changes could always be reverted 

In [156]:
rental_raw_data_parsed = rental_raw_data.copy()

In [157]:
rental_raw_data_parsed['cost_text'].loc[rental_data_cost_text_contains_w_mapped_prices.index] = rental_data_cost_text_contains_w_mapped_prices
rental_raw_data_parsed['cost_text'].loc[cost_text_no_w_but_contain_p.index] = cost_text_no_w_but_contain_p
rental_raw_data_parsed['cost_text'].loc[cost_text_no_w_not_contain_p.index] = cost_text_no_w_not_contain_p

Try to typecast <code>cost_text</code> into a string. Hope there's no errors!

In [158]:
rental_raw_data_parsed['cost_text'] = rental_raw_data_parsed['cost_text'].astype('float')

Brief inspection of unparsed rental price VS parsed rental price

In [159]:
pd.concat([rental_raw_data['cost_text'], rental_raw_data_parsed['cost_text']],axis=1).tail(20)

cost_text  cost_text
14781           $490 per week      490.0
14782                    $450      450.0
14783                    $650      650.0
14784                    $520      520.0
14785                    $520      520.0
14786                    $375      375.0
14787               $448 p.w.      448.0
14788                 $520.00      520.0
14789           $660 per week      660.0
14790           $950 per week      950.0
14791           $460 per week      460.0
14793                    $635      635.0
14794           $640 per week      640.0
14795                    $750      750.0
14796             $525 weekly      525.0
14799                    $400      400.0
14800                    $480      480.0
14801  $440p/w ($1911.90 pcm)      440.0
14802                    $430      430.0
14803                    $380      380.0

# <font color='green'> Delete unnecessesary columns! </font>

In [160]:
rental_raw_data_parsed = rental_raw_data_parsed.drop(columns=['Unnamed: 0', 'desc'])

In [166]:
rental_raw_data_parsed

coordinates  cost_text  \
0             [-37.9032708, 145.0770553]      680.0   
1              [-37.9077745, 145.092738]      650.0   
2             [-37.8922004, 145.0899378]      500.0   
3             [-37.8972564, 145.0871089]      570.0   
4               [-37.894307, 145.082706]      475.0   
...                                  ...        ...   
14799         [-37.0476758, 146.0845677]      400.0   
14800  [-37.05395439999999, 146.0797784]      480.0   
14801         [-37.0551267, 146.0929647]      440.0   
14802         [-37.0561829, 146.0891432]      430.0   
14803         [-37.0494926, 146.0876812]      380.0   

                                          name  \
0       15 Fintonia Street Hughesdale VIC 3166   
1           41 Bishop Street Oakleigh VIC 3166   
2       109/6 Dalgety Street Oakleigh VIC 3166   
3         16/30 Swindon Road Oakleigh VIC 3166   
4      1/14 Bletchley Road Hughesdale VIC 3166   
...                                        ...   
14799       8A Logan Street Mansfield VIC 3722   
14800      27 Elvins Street Mansfield VIC 3722   
14801     1/30 Ailsa Street Mansfield VIC 3722   
14802      49 Hunter Street Mansfield VIC 3722   
14803       6 Clarke Street Mansfield VIC 3722   

                                    rooms  
0      ['3 Beds', '2 Baths', '2 Parking']  
1      ['3 Beds', '2 Baths', '2 Parking']  
2        ['1 Bed', '1 Bath', '1 Parking']  
3      ['3 Beds', '2 Baths', '2 Parking']  
4       ['2 Beds', '1 Bath', '1 Parking']  
...                                   ...  
14799   ['3 Beds', '1 Bath', '2 Parking']  
14800  ['3 Beds', '2 Baths', '4 Parking']  
14801   ['2 Beds', '1 Bath', '3 Parking']  
14802   ['3 Beds', '1 Bath', '3 Parking']  
14803   ['2 Beds', '1 Bath', '2 Parking']  

[14424 rows x 4 columns]

# Cell below is exclusively made for debugging, to make sure that the notebook can be rerun and still produce the same result!

In [162]:
# expected_rental_parsed = pd.read_csv('../data/curated/Rental_ParsedPrices_TwoColsDeleted_Keith.csv')
# for i in range(14424):
#     if rental_raw_data_parsed['cost_text'].iloc[i] != expected_rental_parsed['cost_text'].iloc[i]:
#         print(rental_raw_data_parsed['cost_text'].iloc[i],
#             expected_rental_parsed['cost_text'].iloc[i])

2307.6923076923076 2307.692307692308


In [ ]:
rental_raw_data_parsed.to_csv('../data/curated/All_Houses_Scraped.csv')

In [165]:
ls ../data/curated

Monthly_Median_Suburb_Rents.csv  Rental_ParsedPrices_TwoColsDeleted_Keith.csv
Offence_by_LGA.xls               Schools_by_LGA.xlsx
Preproc_Unemployment_Data.csv    Useful_LGA_Attributes.csv
Preproc_Weather_Data             VIC_SPORTING_FACILITIES.csv
Preproc_Weather_Data.csv         victoria_postcodes.csv
